File: 임현규/programmers/Lv2/올바른_괄호/java/카운트_괄호.java
public class Solution {

    private static final char OPEN_BRACKET = '(';
    private static final char CLOSE_BRACKET = ')';

    public boolean solution(String s) {
        int count = 0;
        for (char ch : s.toCharArray()) {
            count = countOpenBracket(count, ch);
            count = countCloseBracket(count, ch);
            if (isInvalidCount(count)) {
                break;
            }
        }
        return count == 0;
    }

    private int countOpenBracket(int count, char ch) {
        if (ch == OPEN_BRACKET) {
            return count + 1;
        }
        return count;
    }

    private int countCloseBracket(int count, char ch) {
        if (ch == CLOSE_BRACKET) {
            return count - 1;
        }
        return count;
    }

    private boolean isInvalidCount(int count) {
        return count < 0;
    }
}

File: 임현규/programmers/Lv2/올바른_괄호/python/카운트_괄호.py
OPEN_BRACKET = "("
CLOSE_BRACKET = ")"


def solution(s):
    count = 0
    for ch in s:
        count = countOpenBracket(count, ch)
        count = countCloseBracket(count, ch)
        if isInvalidCount(count):
            return False
    return count == 0


def countOpenBracket(count: int, ch: str) -> int:
    if ch == OPEN_BRACKET:
        return count + 1
    return count


def countCloseBracket(count: int, ch: str) -> int:
    if ch == CLOSE_BRACKET:
        return count - 1
    return count


def isInvalidCount(count) -> bool:
    return count < 0
